filegenerator/file_generator.go

```diff
@@ -0,0 +1,368 @@
+				if matched {
+					content, err := ioutil.ReadFile(fullDirPath + "/" + fileName)
+					if err != nil {
+						log.Fatal(err)
+					}
+					for _, line := range strings.Split(string(content), "\n") {
+						if isValidTestLine(line) {
+							testName, stats := parseTestResultLine(line)
+							if _, ok := dirToTestNames[dirName][testName]; !ok {
+								dirToTestNames[dirName][testName] = make(BenchResults)
+							}
+							dirToTestNames[dirName][testName][dataDirName] = stats
+						}
+					}
+				}
+			}
+		}
+	}
+	return dirToTestNames
+}
+
+// GenerateJSONFiles takes a BenchPackages and creates a json file for every test
+// containing its benchmark results data over time.
+func GenerateJSONFiles(packages BenchPackages) {
+	deployRoot := os.Getenv(deployRootEnv)
+	for dir := range packages {
+		for testName := range packages[dir] {
+			testData := packages[dir][testName]
+			fileFullName := deployRoot + "/" + dir + "/" + testName + ".json"
+			testDataJSON, err := json.Marshal(&testData)
+			if err != nil {
+				log.Fatal(err)
+			}
+			err = ioutil.WriteFile(fileFullName, testDataJSON, 0644)
+			if err != nil {
+				log.Fatal(err)
+			}
+		}
+	}
+}
+
+// GenerateTestNameJSONFile creates a JSON file that represents a map of
+// directories to a list of all of the benchmark tests in that directory.
+func GenerateTestNameJSONFile(packages BenchPackages) {
+	deployRoot := os.Getenv(deployRootEnv)
+	testPackageToName := make(map[string][]string)
+	for packageName := range packages {
+		testPackageToName[packageName] = make([]string,
+			len(testPackageToName[packageName]))
+		for testName := range packages[packageName] {
+			testPackageToName[packageName] = append(testPackageToName[packageName], testName)
+		}
+	}
+	fileName := deployRoot + "/test_names.json"
+	testNamesJSON, err := json.Marshal(&testPackageToName)
+	if err != nil {
+		log.Fatal(err)
+	}
+	err = ioutil.WriteFile(fileName, testNamesJSON, 0644)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func getDatesFromPackages(packages BenchPackages) []string {
+	var dates []string
+	dateSet := make(map[string]bool)
+	for _, dir := range packages {
+		for _, tests := range dir {
+			for date := range tests {
+				dateSet[date] = true
+			}
+		}
+	}
+	for date := range dateSet {
+		dates = append(dates, date)
+	}
+	return dates
+}
+
+// GenerateGeometricMeanJSONFile creates a JSON file that has the geometric mean of
+// the results from every benchmark test in a package for every package.
+func GenerateGeometricMeanJSONFile(packages BenchPackages, dirs []string) {
+	deployRoot := os.Getenv(deployRootEnv)
+	dates := getDatesFromPackages(packages)
+	packageToGeometricMean := make(map[string][]GeometricMeanData)
+	for _, dir := range dirs {
+		packageToGeometricMean[dir] = make([]GeometricMeanData, 0)
+		for _, date := range dates {
+			vectors := make(map[string][]float64)
+			testTypes := [4]string{"N", "A", "B", "M"}
+			for _, testType := range testTypes {
```

Comment:
Is this loop necessary? By default, `vectors[anything]` will return nil, and appending to a `nil` slice below is perfectly valid.